lib/std/crypto/argon2.zig

```diff
@@ -0,0 +1,1119 @@
+            &t[i],
+            &t[i + 1],
+            &t[16 + i],
+            &t[16 + i + 1],
+            &t[32 + i],
+            &t[32 + i + 1],
+            &t[48 + i],
+            &t[48 + i + 1],
+            &t[64 + i],
+            &t[64 + i + 1],
+            &t[80 + i],
+            &t[80 + i + 1],
+            &t[96 + i],
+            &t[96 + i + 1],
+            &t[112 + i],
+            &t[112 + i + 1],
+        );
+    }
+    if (xor) {
+        for (t) |v, j| {
+            out[j] ^= in1[j] ^ in2[j] ^ v;
+        }
+    } else {
+        for (t) |v, j| {
+            out[j] = in1[j] ^ in2[j] ^ v;
+        }
+    }
+}
+
+const QuarterRound = struct { a: usize, b: usize, c: usize, d: usize };
+
+fn Rp(a: usize, b: usize, c: usize, d: usize) QuarterRound {
+    return .{ .a = a, .b = b, .c = c, .d = d };
+}
+
+fn fBlaMka(x: u64, y: u64) u64 {
+    const xy = (x & 0xffff_ffff) * (y & 0xffff_ffff);
+    return x +% y +% 2 *% xy;
+}
+
+fn blamkaGeneric(
+    t00: *u64,
+    t01: *u64,
+    t02: *u64,
+    t03: *u64,
+    t04: *u64,
+    t05: *u64,
+    t06: *u64,
+    t07: *u64,
+    t08: *u64,
+    t09: *u64,
+    t10: *u64,
+    t11: *u64,
+    t12: *u64,
+    t13: *u64,
+    t14: *u64,
+    t15: *u64,
+) void {
+    var x = [_]u64{
+        t00.*, t01.*, t02.*, t03.*,
+        t04.*, t05.*, t06.*, t07.*,
+        t08.*, t09.*, t10.*, t11.*,
+        t12.*, t13.*, t14.*, t15.*,
+    };
+    const rounds = comptime [_]QuarterRound{
+        Rp(0, 4, 8, 12),
+        Rp(1, 5, 9, 13),
+        Rp(2, 6, 10, 14),
+        Rp(3, 7, 11, 15),
+        Rp(0, 5, 10, 15),
+        Rp(1, 6, 11, 12),
+        Rp(2, 7, 8, 13),
+        Rp(3, 4, 9, 14),
+    };
+    inline for (rounds) |r| {
+        x[r.a] = fBlaMka(x[r.a], x[r.b]);
+        x[r.d] = math.rotr(u64, x[r.d] ^ x[r.a], 32);
+        x[r.c] = fBlaMka(x[r.c], x[r.d]);
+        x[r.b] = math.rotr(u64, x[r.b] ^ x[r.c], 24);
+        x[r.a] = fBlaMka(x[r.a], x[r.b]);
+        x[r.d] = math.rotr(u64, x[r.d] ^ x[r.a], 16);
+        x[r.c] = fBlaMka(x[r.c], x[r.d]);
+        x[r.b] = math.rotr(u64, x[r.b] ^ x[r.c], 63);
+    }
+    t00.* = x[0];
+    t01.* = x[1];
+    t02.* = x[2];
+    t03.* = x[3];
+    t04.* = x[4];
+    t05.* = x[5];
+    t06.* = x[6];
+    t07.* = x[7];
+    t08.* = x[8];
+    t09.* = x[9];
+    t10.* = x[10];
+    t11.* = x[11];
+    t12.* = x[12];
+    t13.* = x[13];
+    t14.* = x[14];
+    t15.* = x[15];
+}
+
+fn extractKey(
+    blocks: *Blocks,
+    memory: u32,
+    threads: u32,
+    out: []u8,
+    comptime hasher: ?type,
+) void {
+    const lanes = memory / threads;
+    var lane: u32 = 0;
+    while (lane < threads - 1) : (lane += 1) {
+        for (blocks.items[(lane * lanes) + lanes - 1]) |v, i| {
+            blocks.items[memory - 1][i] ^= v;
+        }
+    }
+    var block: [1024]u8 = undefined;
+    for (blocks.items[memory - 1]) |v, i| {
+        mem.writeIntLittle(u64, block[i * 8 ..][0..8], v);
+    }
+    switch (out.len) {
```

Comment:
Same here, use the `expected_out_bits` option.

The output size can be way larger than the block size.

See RFC9106:

```text
           if T <= 64
               H'^T(A) = H^T(LE32(T)||A)
           else
               r = ceil(T/32)-2
               V_1 = H^(64)(LE32(T)||A)
               V_2 = H^(64)(V_1)
               ...
               V_r = H^(64)(V_{r-1})
               V_{r+1} = H^(T-32*r)(V_{r})
               H'^T(X) = W_1 || W_2 || ... || W_r || V_{r+1}
```